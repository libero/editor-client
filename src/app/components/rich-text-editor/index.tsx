import React from 'react';
import { EditorView } from 'prosemirror-view';
import { EditorState, Transaction, TextSelection } from 'prosemirror-state';
import { debounce, get } from 'lodash';
import { DOMSerializer, Slice } from 'prosemirror-model';
import 'prosemirror-example-setup/style/style.css';
import 'prosemirror-menu/style/menu.css';
import { SectionContainer, SectionContainerVariant } from 'app/components/section-container';
import { ReferenceCitationNodeView } from 'app/components/reference-citation-editor-popup';
import { ComponentWithId } from 'app/utils/types';
import { BoxTextNodeView } from 'app/components/box-text';
import { hasParentNodeOf } from 'app/utils/view.utils';
import { LinkNodeView } from 'app/components/link-editor-popup';
import { FigureNodeView } from 'app/components/figure';
import { FigureCitationNodeView } from 'app/components/figure-citation';

/*
 * available classes:
 *   - root: applied to SectionContainer
 * */

export interface RichTextEditorProps {
  editorState: EditorState;
  label?: string;
  classes?: Record<string, string>;
  name?: string;
  isActive: boolean;
  variant?: SectionContainerVariant;
  onChange?: (change: Transaction, name: string) => void;
  onFocus?: (state: EditorState, name: string) => void;
  onBlur?: (state: EditorState, name: string) => void;
}

const restoreSelection = debounce((editorView, from, to) => {
  if (editorView.state.selection) {
    const $from = editorView.state.doc.resolve(from);
    const $to = editorView.state.doc.resolve(to);
    const change = editorView.state.tr.setSelection(new TextSelection($from, $to));
    editorView.dispatch(change);
  }
}, 50);

export class RichTextEditor extends React.Component<ComponentWithId<RichTextEditorProps>> {
  private options;
  public editorView: EditorView;

  constructor(props: ComponentWithId<RichTextEditorProps>) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.options = {
      nodeViews: {
        refCitation(node, view, getPos) {
          return new ReferenceCitationNodeView(node, view, getPos);
        },
        boxText: (node, view, getPos) => {
          return new BoxTextNodeView(node, view, getPos);
        },
        link(node, view) {
          return new LinkNodeView(node, view);
        },
        figure: (node, view, getPos) => {
          return new FigureNodeView(node, view, getPos);
        },
        figureCitation: (node, view, getPos) => {
          return new FigureCitationNodeView(node, view, getPos);
        }
      },
      handleDOMEvents: {
        focus: ({ state }: EditorView) => {
          if (this.props.onFocus && !this.props.isActive) {
            this.props.onFocus(state, this.props.name);
          }
          return true;
        },
        blur: ({ state }: EditorView) => {
          if (this.props.onBlur && this.props.isActive) {
            this.props.onBlur(state, this.props.name);
          }
          return true;
        }
      }
    };
  }

  focus() {
    this.focusEditor();
  }

  blur() {
    (this.editorView.dom as HTMLDivElement).blur();
  }

  shouldComponentUpdate(nextProps: RichTextEditorProps) {
    // editor state is kept in sync between app state and editor state. App state will change when using formatting
    // or any toolbar menu. In this case when change comes from outside it needs to override editor state
    if (!nextProps.editorState.doc.eq(this.props.editorState.doc)) {
      this.updateEditorState(nextProps.editorState);
    }

    return (
      this.props.label !== nextProps.label ||
      this.props.isActive !== nextProps.isActive ||
      this.props.variant !== nextProps.variant
    );
  }

  componentDidUpdate() {
    // when component updates we need to restore focus
    this.restoreFocus();
  }

  componentWillUnmount() {
    if (this.editorView) {
      this.editorView.destroy();
    }
  }

  updateEditorState(editorState: EditorState) {
    if (editorState !== this.editorView.state) {
      this.editorView.updateState(editorState);
    }
  }

  render() {
    return (
      <SectionContainer
        label={this.props.label}
        className={get(this.props, 'classes.root', '')}
        focused={this.props.isActive}
        id={this.props.id}
        variant={this.props.variant}
      >
        <div ref={this.createEditorView} className="prosemirrorContainer" />
      </SectionContainer>
    );
  }

  private async handleChange(change: Transaction) {
    const newState = this.editorView.state.apply(change);
    this.updateEditorState(newState);
    this.props.onChange(change, this.props.name);
  }

  private createEditorView = (element: HTMLElement) => {
    if (element) {
      const clipboardSerializer = DOMSerializer.fromSchema(this.props.editorState.schema);
      Object.entries(this.props.editorState.schema.nodes).forEach(([nodeName, nodeType]) => {
        clipboardSerializer.nodes[nodeName] = get(nodeType, 'spec.toClipboardDOM', clipboardSerializer.nodes[nodeName]);
      });

      const additionalOptions = this.options || {};
      this.editorView = new EditorView(element, {
        ...additionalOptions,
        clipboardSerializer,
        handleScrollToSelection: (view: EditorView) => {
          return this.isFocusControlDelegated();
        },
        clipboardTextSerializer,
        state: this.props.editorState,
        dispatchTransaction: this.handleChange
      });
    }
  };

  private focusEditor() {
    const { from, to } = this.editorView.state.selection;
    restoreSelection(this.editorView, from, to);
    this.editorView.focus();
  }

  private isFocusControlDelegated() {
    const { $from } = this.editorView.state.selection;
    return this.editorView.dom.contains(document.activeElement) || hasParentNodeOf($from, ['boxText', 'figure']);
  }

  // focus is restored based on 3 conditions
  // #1 - component is active acc. to application state (prop: isActive)
  // #2 - editor does not have focus
  // #3 - focus is not delegated to a node view
  private restoreFocus() {
    if (this.props.isActive && this.editorView && !this.editorView.hasFocus()) {
      if (!this.isFocusControlDelegated()) {
        this.focusEditor();
      }
    }
  }
}

function clipboardTextSerializer(slice: Slice): string {
  const content = slice.content;
  const text = [];
  let separated = true;
  const blockSeparator = ' ';
  content.nodesBetween(0, content.size, (node, pos) => {
    if (node.isText) {
      text.push(node.text);
      separated = !blockSeparator;
    } else if (node.isLeaf && node.type.spec.toClipboardText) {
      text.push(node.type.spec.toClipboardText(node));
      separated = !blockSeparator;
    } else if (!separated && node.isBlock) {
      text.push(blockSeparator);
      separated = true;
    }
    return true;
  });

  return text.join('');
}
