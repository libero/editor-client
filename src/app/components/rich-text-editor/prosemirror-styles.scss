@import '../../styles/constants.module';

.prosemirrorContainer {
  line-height: initial;
  color: $text-primary;

  .ProseMirror .placeholder {
    color: rgba(0,0,0,0.38);
    font-weight: normal;
    pointer-events: none;
    height: 0;
  }

  .ProseMirror:focus .placeholder {
    display: none;
  }

  .selection-marker {
    background: #338FFF;
    color: white;
  }

  .ProseMirror {
    ::selection {
      background: transparent;
    }

    outline: none;

    h1 {
      font-size: 26px;
      font-weight: 600;
      margin: 40px 0 16px 0;
      line-height: 1.2;
    }
    h2 {
      font-size: 24px;
      font-weight: 600;
      margin: 40px 0 16px 0;
      line-height: 1.2;
    }
    h3 {
      font-size: 22px;
      font-weight: 600;
      margin: 40px 0 16px 0;
      line-height: 1.2;
    }
    h4 {
      font-size: 20px;
      font-weight: 600;
      margin: 32px 0 16px 0;
      line-height: 1.2;
    }
    h5 {
      font-size: 18px;
      font-weight: 600;
      margin: 32px 0 16px 0;
      line-height: 1.2;
    }
    h6 {
      font-size: 16px;
      font-weight: 600;
      margin: 32px 0 16px 0;
      line-height: 1.2;
    }
    p {
      font-size: 16px;
      line-height: 1.5;
      margin: 0 0 16px 0;
      &:last-child {
        margin-bottom: 0;
      }
    }

    h1.article-title {
      margin: 0;
      padding: 0;
    }
  }
}
