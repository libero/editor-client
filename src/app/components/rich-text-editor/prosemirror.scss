.ProseMirror {
  .article-title {
    font-size: 36px;
    font-weight: 500;
  }
}