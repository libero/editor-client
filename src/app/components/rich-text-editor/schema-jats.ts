/* eslint-disable */
import { DOMOutputSpec, Schema, SchemaSpec } from 'prosemirror-model';

const pDOM = ['p', 0] as DOMOutputSpec,
  hrDOM = ['hr'] as DOMOutputSpec,
  preDOM = ['pre', ['code', 0]] as DOMOutputSpec,
  brDOM = ['br'] as DOMOutputSpec;

// :: Object
// [Specs](#model.NodeSpec) for the nodes defined in this schema.
export const nodes = {
  // :: NodeSpec The top level document node.
  doc: {
    content: 'block+'
  },

  article: {
    content: 'body*',
    parseDOM: [{ tag: 'article' }],
    toDOM() {
      return ['div'] as DOMOutputSpec;
    }
  },

  body: {
    content: 'block*',
    parseDOM: [{ tag: 'body' }],
    toDOM() {
      return ['div', 0] as DOMOutputSpec;
    }
  },

  boxed_text: {
    content: 'paragraph*',
    group: 'block',
    parseDOM: [{ tag: 'boxed-text' }],
    toDOM() {
      return ['div', { class: 'sc-card' }, 0] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A plain paragraph textblock. Represented in the DOM
  // as a `<p>` element.
  paragraph: {
    content: 'inline*',
    group: 'block',
    parseDOM: [{ tag: 'p' }],
    toDOM() {
      return pDOM;
    }
  },

  // :: NodeSpec A blockquote (`<blockquote>`) wrapping one or more blocks.
  blockquote: {
    content: 'block+',
    group: 'block',
    defining: true,
    parseDOM: [{ tag: 'blockquote' }],
    toDOM() {
      return ['div'] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A horizontal rule (`<hr>`).
  horizontal_rule: {
    group: 'block',
    parseDOM: [{ tag: 'hr' }],
    toDOM() {
      return hrDOM;
    }
  },

  // :: NodeSpec A heading textblock, with a `level` attribute that
  // should hold the number 1 to 6. Parsed and serialized as `<h1>` to
  // `<h6>` elements.
  heading: {
    attrs: { level: { default: 1 } },
    content: 'inline*',
    group: 'block',
    defining: true,
    parseDOM: [
      { tag: 'h1', attrs: { level: 1 } },
      { tag: 'h2', attrs: { level: 2 } },
      { tag: 'h3', attrs: { level: 3 } },
      { tag: 'h4', attrs: { level: 4 } },
      { tag: 'h5', attrs: { level: 5 } },
      { tag: 'h6', attrs: { level: 6 } }
    ],
    toDOM(node) {
      return ['h' + node.attrs.level, 0] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A code listing. Disallows marks or non-text inline
  // nodes by default. Represented as a `<pre>` element with a
  // `<code>` element inside of it.
  code_block: {
    content: 'text*',
    marks: '',
    group: 'block',
    code: true,
    defining: true,
    parseDOM: [{ tag: 'pre', preserveWhitespace: 'full' }],
    toDOM() {
      return preDOM;
    }
  },

  // :: NodeSpec The text node.
  text: {
    group: 'inline'
  },

  // :: NodeSpec An inline image (`<img>`) node. Supports `src`,
  // `alt`, and `href` attributes. The latter two default to the empty
  // string.
  image: {
    inline: true,
    attrs: {
      src: {},
      alt: { default: null },
      title: { default: null }
    },
    group: 'inline',
    draggable: true,
    parseDOM: [
      {
        tag: 'img[src]',
        getAttrs(dom) {
          return {
            src: dom.getAttribute('src'),
            title: dom.getAttribute('title'),
            alt: dom.getAttribute('alt')
          };
        }
      }
    ],
    toDOM(node) {
      const { src, alt, title } = node.attrs;
      return ['img', { src, alt, title }] as DOMOutputSpec;
    }
  },

  // :: NodeSpec A hard line break, represented in the DOM as `<br>`.
  hard_break: {
    inline: true,
    group: 'inline',
    selectable: false,
    parseDOM: [{ tag: 'br' }],
    toDOM() {
      return brDOM;
    }
  },

  ext_link: {
    inline: true,
    content: 'inline*',
    attrs: {
      type: { default: 'uri' },
      href: {}
    },
    group: 'inline',
    parseDOM: [
      {
        tag: 'ext-link',
        getAttrs(dom) {
          return {
            type: dom.getAttribute('ext-link-type'),
            href: dom.getAttribute('xlink:href')
          };
        }
      }
    ],
    toDOM(node) {
      const { href } = node.attrs;
      return ['a', { href }, 0] as DOMOutputSpec;
    }
  },

  //<xref ref-type="bibr" rid="bib5">Katz et al., 2009</xref>
  xref: {
    inline: true,
    content: 'inline*',
    attrs: {
      type: { default: 'bibr' },
      href: { default: '#' }
    },
    group: 'inline',
    parseDOM: [
      {
        tag: 'xref',
        getAttrs(dom) {
          return {
            type: dom.getAttribute('ref-type'),
            href: dom.getAttribute('rid')
          };
        }
      }
    ],
    toDOM(node) {
      const { href } = node.attrs;
      return ['a', { href }, 0] as DOMOutputSpec;
    }
  },

  figure: {
    content: 'label* caption* graphic* attrib*',
    group: 'block',
    attrs: {
      id: {},
      position: { default: 'float' }
    },
    parseDOM: [
      {
        tag: 'fig',
        getAttrs(dom) {
          return {
            id: dom.getAttribute('id'),
            position: dom.getAttribute('position')
          };
        }
      }
    ],
    toDOM(node) {
      return ['div', { class: 'sc-card' }, 0] as DOMOutputSpec;
    }
  },

  label: {
    content: 'inline*',
    group: 'figure',
    selectable: false,
    parseDOM: [
      {
        tag: 'label'
      }
    ],
    toDOM(node) {
      return ['h2', 0] as DOMOutputSpec;
    }
  },

  caption: {
    content: 'title* paragraph*',
    group: 'figure',
    parseDOM: [
      {
        tag: 'caption'
      }
    ],
    toDOM() {
      return ['div', 0] as DOMOutputSpec;
    }
  },

  title: {
    content: 'text*',
    parseDOM: [
      {
        tag: 'title'
      }
    ],
    toDOM() {
      return ['h3', 0] as DOMOutputSpec;
    }
  },

  graphic: {
    attrs: {
      src: {}
    },
    group: 'figure',
    draggable: true,
    parseDOM: [
      {
        tag: 'graphic',
        getAttrs(dom) {
          return {
            src: dom.getAttribute('xlink:href')
          };
        }
      }
    ],
    toDOM(node) {
      const { src } = node.attrs;
      return ['img', { src }] as DOMOutputSpec;
    }
  },

  attrib: {
    content: 'text*',
    parseDOM: [
      {
        tag: 'attrib'
      }
    ],
    toDOM(node) {
      return ['h3', 0] as DOMOutputSpec;
    }
  }
};

const emDOM = ['em', 0] as DOMOutputSpec,
  strongDOM = ['strong', 0] as DOMOutputSpec,
  codeDOM = ['code', 0] as DOMOutputSpec;

// :: Object [Specs](#model.MarkSpec) for the marks in the schema.
export const marks = {
  // :: MarkSpec A link. Has `href` and `title` attributes. `title`
  // defaults to the empty string. Rendered and parsed as an `<a>`
  // element.
  link: {
    attrs: {
      href: {},
      title: { default: null }
    },
    inclusive: false,
    parseDOM: [
      {
        tag: 'a[href]',
        getAttrs(dom) {
          return { href: dom.getAttribute('href'), title: dom.getAttribute('title') };
        }
      }
    ],
    toDOM(node) {
      const { href, title } = node.attrs;
      return ['a', { href, title }, 0] as DOMOutputSpec;
    }
  },

  // :: MarkSpec An emphasis mark. Rendered as an `<em>` element.
  // Has parse rules that also match `<i>` and `font-style: italic`.
  em: {
    parseDOM: [{ tag: 'i' }, { tag: 'italic' }, { tag: 'em' }, { style: 'font-style=italic' }],
    toDOM() {
      return emDOM;
    }
  },

  // :: MarkSpec A strong mark. Rendered as `<strong>`, parse rules
  // also match `<b>` and `font-weight: bold`.
  strong: {
    parseDOM: [
      { tag: 'strong' },
      { tag: 'bold' },
      // This works around a Google Docs misbehavior where
      // pasted content will be inexplicably wrapped in `<b>`
      // tags with a font-weight normal.
      { tag: 'b', getAttrs: (node) => node.style.fontWeight !== 'normal' && null },
      { style: 'font-weight', getAttrs: (value) => /^(bold(er)?|[5-9]\d{2,})$/.test(value) && null }
    ],
    toDOM() {
      return strongDOM;
    }
  },

  // :: MarkSpec Code font mark. Represented as a `<code>` element.
  code: {
    parseDOM: [{ tag: 'code' }],
    toDOM() {
      return codeDOM;
    }
  }
};

// :: Schema
// This schema roughly corresponds to the document schema used by
// [CommonMark](http://commonmark.org/), minus the list elements,
// which are defined in the [`prosemirror-schema-list`](#schema-list)
// module.
//
// To reuse elements from this schema, extend or read from its
// `spec.nodes` and `spec.marks` [properties](#model.Schema.spec).

type NodesType = keyof typeof nodes;
type MarksType = keyof typeof marks;
export const schema = new Schema<NodesType, MarksType>({ nodes, marks } as SchemaSpec<NodesType, MarksType>);
/* eslint-enable */
