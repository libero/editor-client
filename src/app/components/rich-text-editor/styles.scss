.editorview-wrapper {
  width: 100%;
  height: 100%;
  position: relative;
  padding: 8px;
  border: solid 1px rgba(0, 0, 0, 0.23);
  border-radius: 4px;

  .editor-label {
    font-size: 12px;
    color: #263238;
    padding: 0 4px;
  }
}
