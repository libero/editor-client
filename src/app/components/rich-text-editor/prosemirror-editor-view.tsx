import React from 'react';
import { EditorState, Transaction } from 'prosemirror-state';
import { EditorProps, EditorView } from 'prosemirror-view';
import classnames from 'classnames';
import 'prosemirror-view/style/prosemirror.css';

import './prosemirror-styles.scss';

interface ProseMirrorEditorViewState {
  editorState: EditorState;
}

export interface ProseMirrorEditorViewProps {
  className?: string;
  editorState: EditorState;
  options?: Partial<EditorProps>;
  onChange: (tx: Transaction) => void;
}

export class ProseMirrorEditorView extends React.Component<ProseMirrorEditorViewProps, ProseMirrorEditorViewState> {
  public props;
  public editorView: EditorView;

  focus() {
    // debugger;
    this.editorView.focus();
  }

  blur() {
    (this.editorView.dom as HTMLDivElement).blur();
  }

  shouldComponentUpdate(nextProps) {
    if (nextProps.editorState !== this.editorView.state) {
      this.editorView.updateState(nextProps.editorState);
    }
    return false;
  }

  componentWillUnmount() {
    if (this.editorView) {
      this.editorView.destroy();
    }
  }

  render() {
    // Render just an empty div which is then used as a container for an
    // EditorView instance.
    return <div ref={this.createEditorView} className={classnames('prosemirrorContainer', this.props.className)} />;
  }

  private createEditorView = (element: HTMLElement) => {
    if (element) {
      const additionalOptions = this.props.options || {};
      this.editorView = new EditorView(element, {
        ...additionalOptions,
        state: this.props.editorState,
        dispatchTransaction: (tx: Transaction) => this.props.onChange(tx)
      });
    }
  };
}
