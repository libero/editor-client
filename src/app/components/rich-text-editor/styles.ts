import { makeStyles } from '@material-ui/core/styles';

export const useRichTextEditorStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    height: '100%',
    padding: theme.spacing(1, 1.5, 1.5, 1.5),
    border: `solid 1px`,
    borderColor: theme.palette.secondary.dark,
    borderRadius: theme.spacing(0.5),
    boxSizing: 'border-box'
  },

  label: {
    fontSize: '0.75rem',
    color: theme.palette.text.secondary,
    padding: theme.spacing(0, 0.5)
  },

  focused: {
    borderColor: theme.palette.primary.light,
    '& > legend': {
      color: theme.palette.primary.light
    }
  }
}));
