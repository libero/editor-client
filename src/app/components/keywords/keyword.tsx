import { EditorState, Transaction } from 'prosemirror-state';
import React, { useCallback, useEffect, useRef } from 'react';
import classNames from 'classnames';
import { EditorView } from 'prosemirror-view';
import { IconButton } from '@material-ui/core';
import CancelIcon from '@material-ui/icons/Cancel';

import { ProseMirrorEditorView } from 'app/components/rich-text-editor/prosemirror-editor-view';
import { useKeywordStyles } from './styles';

const ENTER_KEY_CODE = 'Enter';

interface KeywordProps {
  editorState: EditorState;
  onChange: (state: Transaction) => void;
  onDelete: () => void;
  isActive: boolean;
  onFocus: (state: EditorState) => void;
  onBlur: () => void;
}

export const Keyword: React.FC<KeywordProps> = (props) => {
  const { editorState, onDelete, onChange, onFocus, isActive, onBlur } = props;
  const prosemirrorRef = useRef(null);
  const containerRef = useRef<HTMLDivElement>(null);
  const deleteButtonRef = useRef(null);
  const classes = useKeywordStyles();

  const focusOnDblClick = useCallback(() => {
    prosemirrorRef.current.focus();
  }, []);

  const preventSingleClickWhenInactive = useCallback(
    (event: Event) => {
      const isDeleteButtonClick = deleteButtonRef.current.contains(event.target);
      if (!isActive) {
        event.stopPropagation();
        event.preventDefault();
      }

      if (isDeleteButtonClick) {
        onDelete();
      }
    },
    [isActive, onDelete]
  );

  useEffect(() => {
    const containerEl = containerRef.current;
    if (containerEl) {
      containerEl.addEventListener('dblclick', focusOnDblClick, true);
      containerEl.addEventListener('mousedown', preventSingleClickWhenInactive, true);
    }

    return () => {
      containerEl.removeEventListener('dblclick', focusOnDblClick, true);
      containerEl.removeEventListener('mousedown', preventSingleClickWhenInactive, true);
    };
  }, [containerRef, focusOnDblClick, preventSingleClickWhenInactive]);

  const handleFocusEvent = (view: EditorView): boolean => {
    onFocus(view.state);
    return true;
  };

  const options = {
    handleDOMEvents: {
      focus: handleFocusEvent,
      keydown: (view: EditorView, event: KeyboardEvent): boolean => {
        if (event.key === ENTER_KEY_CODE) {
          prosemirrorRef.current.blur();
          onBlur();
          return true;
        }
        return false;
      }
    }
  };

  return (
    <div
      className={classNames(classes.keyword, { focused: isActive })}
      ref={containerRef}
      data-test-id="keyword-container"
    >
      <ProseMirrorEditorView options={options} ref={prosemirrorRef} editorState={editorState} onChange={onChange} />
      <IconButton
        ref={deleteButtonRef}
        aria-label="delete keyword"
        tabIndex={0}
        color="primary"
        size="small"
        className={isActive ? classes.hidden : ''}
      >
        <CancelIcon className={classes.deleteIcon} />
      </IconButton>
    </div>
  );
};
