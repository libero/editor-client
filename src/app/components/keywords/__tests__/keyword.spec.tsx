import React from 'react';
import { EditorState } from 'prosemirror-state';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

import { Keyword } from '../keyword';
import { ProseMirrorEditorView } from '../../rich-text-editor/prosemirror-editor-view';
import { createNewKeywordState } from '../../../models/manuscript-state.factory';

describe('KeywordsEditorComponent', () => {
  it('renders keyword', () => {
    const props = {
      editorState: new EditorState(),
      onChange: jest.fn(),
      onDelete: jest.fn()
    };

    const component = create(<Keyword {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('sets focus on dbl click', () => {
    const props = {
      editorState: createNewKeywordState(),
      onChange: jest.fn(),
      onDelete: jest.fn()
    };

    const component = mount(<Keyword {...props} />);
    const prosemirror = component.find(ProseMirrorEditorView).get(0);
    jest.spyOn(prosemirror['ref'].current, 'focus');

    component.find('.keyword').getDOMNode().dispatchEvent(new Event('dblclick'));

    expect(prosemirror['ref'].current.focus).toHaveBeenCalled();
  });

  it('stops single clicks', () => {
    const props = {
      editorState: createNewKeywordState(),
      onChange: jest.fn(),
      onDelete: jest.fn()
    };

    const component = mount(<Keyword {...props} />);
    const prosemirror = component.find(ProseMirrorEditorView).get(0);

    jest.spyOn(prosemirror['ref'].current, 'focus');
    const mouseDownEvent = new Event('mousedown');
    jest.spyOn(mouseDownEvent, 'preventDefault');
    jest.spyOn(mouseDownEvent, 'stopPropagation');
    component.find('.keyword').getDOMNode().dispatchEvent(mouseDownEvent);

    expect(mouseDownEvent.preventDefault).toHaveBeenCalled();
    expect(mouseDownEvent.stopPropagation).toHaveBeenCalled();
  });

  it('should unsubscribe on unmount', () => {
    const props = {
      editorState: createNewKeywordState(),
      onChange: jest.fn(),
      onDelete: jest.fn()
    };

    const component = mount(<Keyword {...props} />);
    const containerDiv = component.find('.keyword').getDOMNode();
    jest.spyOn(containerDiv, 'removeEventListener');
    component.unmount();
    expect(containerDiv.removeEventListener).toHaveBeenCalledWith('dblclick', expect.any(Function), true);
    expect(containerDiv.removeEventListener).toHaveBeenCalledWith('mousedown', expect.any(Function), true);
  });
});
