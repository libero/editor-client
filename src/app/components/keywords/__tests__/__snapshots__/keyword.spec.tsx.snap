// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeywordsEditorComponent renders keyword 1`] = `
<div
  className="makeStyles-keyword-1"
  data-test-id="keyword-container"
>
  <div
    className="prosemirrorContainer"
  />
  <button
    aria-label="delete keyword"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-deleteIcon-3"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
    </span>
  </button>
</div>
`;
